# <center>Pydata Yerevan 2022 : How to use Pandas Efficiently </center>

### Github Repository :  <a href="https://bit.ly/3QeAJ3e">bit.ly/3QeAJ3e</a>
### Google Colab :  <a href ="http://bit.ly/3dgwqpr"> bit.ly/3dgwqpr </a>
### Survey:

While you are waiting for the tutorial to start, you can go through above link to download dataset and Notebook :<br>
N.B. : If you do not have Python installed on your system, you can use Google Colab to run the notebook.

We will be using the following dataset for this tutorial :<br>
- IMDB top 1000 Movie Dataset<br> https://www.kaggle.com/datasets/mysarahmadbhat/imdb-top-1000-movies
- Bank Marketing Dataset (UCI) https://archive.ics.uci.edu/ml/datasets/bank+marketing <br>
- Bike Sharing dataset (Kaggle) https://www.kaggle.com/datasets/yasserh/bike-sharing-dataset



### PART 1 :Pandas Big Picture

What is Pandas?<br>


<img src="img/Excel_Vs_Pandas.jpg" width="550" height="200" style ="float:center" >

### Why pandas? 
-  easy to use 
-  easy to learn.
- it is fast 
- it is a powerful data analysis toolkit
- it is open source
- it is supported by many developers (~3k)


### Two building blocks of pandas are Series and DataFrame
- Series: A 1D labeled data structure with rows.
- DataFrame: A 2D labeled data structure with columns and rows.


In [2]:
#import required libraries
import pandas as pd
import numpy as np


In [2]:
# read top 1000 imdb movie dataset of the dataset
imdb = pd.read_csv('data/imdb_1000.csv')
imdb.head()


Name  Year  duration  Rating  Meta_score    Votes  \
0  The Shawshank Redemption  1994       142     9.3        80.0  2377480   
1             The Godfather  1972       175     9.2       100.0  1646818   
2           The Dark Knight  2008       152     9.0        84.0  2341261   
3    The Godfather: Part II  1974       202     9.0        90.0  1146666   
4              12 Angry Men  1957        96     9.0        96.0   701882   

               Director                                               cast  \
0        Frank Darabont  Tim Robbins, Morgan Freeman, Bob Gunton, Willi...   
1  Francis Ford Coppola  Marlon Brando, Al Pacino, James Caan, Diane Ke...   
2     Christopher Nolan  Christian Bale, Heath Ledger, Aaron Eckhart, M...   
3  Francis Ford Coppola  Al Pacino, Robert De Niro, Robert Duvall, Dian...   
4          Sidney Lumet  Henry Fonda, Lee J. Cobb, Martin Balsam, John ...   

    Gross   genre  
0   28.34   Drama  
1  134.97   Crime  
2  534.86  Action  
3   57.30   Crime  
4    4.36   Crime

In [6]:
#Select Rows
imdb.loc[:,:]


Name  Year  duration  Rating  Meta_score    Votes  \
0    The Shawshank Redemption  1994       142     9.3        80.0  2377480   
1               The Godfather  1972       175     9.2       100.0  1646818   
2             The Dark Knight  2008       152     9.0        84.0  2341261   
3      The Godfather: Part II  1974       202     9.0        90.0  1146666   
4                12 Angry Men  1957        96     9.0        96.0   701882   
..                        ...   ...       ...     ...         ...      ...   
995    Breakfast at Tiffany's  1961       115     7.6        76.0   168713   
996                     Giant  1956       201     7.6        84.0    34594   
997     From Here to Eternity  1953       118     7.6        85.0    43868   
998                  Lifeboat  1944        97     7.6        78.0    26791   
999              The 39 Steps  1935        86     7.6        93.0    52688   

                 Director                                               cast  \
0          Frank Darabont  Tim Robbins, Morgan Freeman, Bob Gunton, Willi...   
1    Francis Ford Coppola  Marlon Brando, Al Pacino, James Caan, Diane Ke...   
2       Christopher Nolan  Christian Bale, Heath Ledger, Aaron Eckhart, M...   
3    Francis Ford Coppola  Al Pacino, Robert De Niro, Robert Duvall, Dian...   
4            Sidney Lumet  Henry Fonda, Lee J. Cobb, Martin Balsam, John ...   
..                    ...                                                ...   
995         Blake Edwards  Audrey Hepburn, George Peppard, Patricia Neal,...   
996        George Stevens  Elizabeth Taylor, Rock Hudson, James Dean, Car...   
997        Fred Zinnemann  Burt Lancaster, Montgomery Clift, Deborah Kerr...   
998      Alfred Hitchcock  Tallulah Bankhead, John Hodiak, Walter Slezak,...   
999      Alfred Hitchcock  Robert Donat, Madeleine Carroll, Lucie Mannhei...   

      Gross   genre  
0     28.34   Drama  
1    134.97   Crime  
2    534.86  Action  
3     57.30   Crime  
4      4.36   Crime  
..      ...     ...  
995     NaN  Comedy  
996     NaN   Drama  
997   30.50   Drama  
998     NaN   Drama  
999     NaN   Crime  

[1000 rows x 10 columns]

In [7]:
#Select Columns
imdb.loc[:,['Name','Year']]



Name  Year
0    The Shawshank Redemption  1994
1               The Godfather  1972
2             The Dark Knight  2008
3      The Godfather: Part II  1974
4                12 Angry Men  1957
..                        ...   ...
995    Breakfast at Tiffany's  1961
996                     Giant  1956
997     From Here to Eternity  1953
998                  Lifeboat  1944
999              The 39 Steps  1935

[1000 rows x 2 columns]

In [12]:
# Select with iloc
imdb.iloc[1:6,:]




Name  Year  duration  Rating  \
1                                  The Godfather  1972       175     9.2   
2                                The Dark Knight  2008       152     9.0   
3                         The Godfather: Part II  1974       202     9.0   
4                                   12 Angry Men  1957        96     9.0   
5  The Lord of the Rings: The Return of the King  2003       201     8.9   

   Meta_score    Votes              Director  \
1       100.0  1646818  Francis Ford Coppola   
2        84.0  2341261     Christopher Nolan   
3        90.0  1146666  Francis Ford Coppola   
4        96.0   701882          Sidney Lumet   
5        94.0  1663902         Peter Jackson   

                                                cast   Gross   genre  
1  Marlon Brando, Al Pacino, James Caan, Diane Ke...  134.97   Crime  
2  Christian Bale, Heath Ledger, Aaron Eckhart, M...  534.86  Action  
3  Al Pacino, Robert De Niro, Robert Duvall, Dian...   57.30   Crime  
4  Henry Fonda, Lee J. Cobb, Martin Balsam, John ...    4.36   Crime  
5  Elijah Wood, Viggo Mortensen, Ian McKellen, Or...  377.85  Action

In [13]:
# Use np.where 
imdb['has_high_rate'] = np.where(imdb.Rating > 8.0 ,'High','low')
imdb.head()



Name  Year  duration  Rating  Meta_score    Votes  \
0  The Shawshank Redemption  1994       142     9.3        80.0  2377480   
1             The Godfather  1972       175     9.2       100.0  1646818   
2           The Dark Knight  2008       152     9.0        84.0  2341261   
3    The Godfather: Part II  1974       202     9.0        90.0  1146666   
4              12 Angry Men  1957        96     9.0        96.0   701882   

               Director                                               cast  \
0        Frank Darabont  Tim Robbins, Morgan Freeman, Bob Gunton, Willi...   
1  Francis Ford Coppola  Marlon Brando, Al Pacino, James Caan, Diane Ke...   
2     Christopher Nolan  Christian Bale, Heath Ledger, Aaron Eckhart, M...   
3  Francis Ford Coppola  Al Pacino, Robert De Niro, Robert Duvall, Dian...   
4          Sidney Lumet  Henry Fonda, Lee J. Cobb, Martin Balsam, John ...   

    Gross   genre has_high_rate  
0   28.34   Drama          High  
1  134.97   Crime          High  
2  534.86  Action          High  
3   57.30   Crime          High  
4    4.36   Crime          High

In [16]:
# use Value_Counts to counts number of rows
imdb.has_high_rate.value_counts(normalize=True) *100


low     67.2
High    32.8
Name: has_high_rate, dtype: float64

In [17]:
# check the type of the dataset / columns 
type(imdb)


pandas.core.frame.DataFrame

In [18]:
# check the type of a column
type(imdb.Name)


pandas.core.series.Series

### some of the basic operation by Pandas


In [ ]:
# print some basic information about the dataset
imdb.Rating.min()


### Notes:
#### 1- Ask Questions and Answer them with Data by Pandas insetad of Jumping to the analysis without knowing the data  <br>
#### 2- Relation does not mean causation. Here we just find the relationship is present in the data.(Causation needs more domain knowledge)  <br>
#### 3- There more than one way to do something in Pandas. <br>
#### 4- Data is not always clean. We need to clean it before analysis. <br>
Q1 : What is the trend of Years in which movies are released? <br>
Q2 : Which Genre had the highest number of movies among top 1000 movies dataset? <br>
Q3 : What is the distribution of Duration (In every 10 minutes) in the top 1000 movies? <br>
Q4 : what is the average duration of each genre?    <br>
Q5 : Was there an increase of Votes over the time ? <br
Q6 : For each year what was the average rating and the maximum of Meta score? <br>


#### Practice 1:
How many of movies got rating between 7.5 and 8.5? (Hint:use value_counts() method) <br>
What is trend of Gross Earning over the years? (Hint: use groupby() method) <br>
what is the distibution of meta score in every 10 points? (Hint: use cut() method) <br>
How many cast are there in each movei? (Hint: use apply() method) <br>

In [ ]:
#What is the trend of Years in which movies are released?


In [ ]:
#Which Genre had the highest number of movies among top 1000 movies dataset? 


In [ ]:
# What is the distribution of Duration (In every 10 minutes) in the top 1000 movies?
# way 1:
bins = [60,70,80,90,100,110,120,130,140,150,160,170,180,190,200,210,220,230,240,250]
labels = ['60-70','70-80','80-90','90-100','100-110','110-120','120-130','130-140','140-150','150-160','160-170','170-180','180-190','190-200','200-210','210-220','220-230','230-240','240-250']


In [ ]:
# way 2: using math function


In [ ]:
# what is the average duration of each genre?


In [ ]:
# Was there an increase of Votes over the time ? 


In [ ]:
# For each year what was the average rating and the maximum of Meta score?


In [ ]:
# For each year what was the average rating and the maximum of Meta score?(with Pivot table) 


### Part II : How to use Pandas More effiently

### 95% of people use this kind of code style but only 5% of people know how to use it more efficiently.!

<img src="img/Screenshot1.jpg" width="750" height="500" style ="float:middle" ><br>
<br>
<img src="img/Screenshot2.jpg" width="750" height="500" style ="float:left" >


### Use chaining instead of classic coding
 ### what is Chainning?
 Chaining allows you to express the idea of data flowing through a pipeline almost 1-to-1 into your code, making your code more readable both for yourself and for others. <br>
<br>
<img src="img/Chaining.png" width="750" height="200"> <br>
*source :www.aaa-riskfinance.es/publicaties/chaining-in-python-and-r/


#### 1- Benefit of usnig chaining in your codes
- It is more readable and understandable
- It can save you a lot of time
- It is more pythonic 
- It is more fun!


In [3]:
# read bank dataset



nput variables:<br>
**bank client data:**<br>
1 - age (numeric) <br>
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')<br>
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)<br>
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')<br>
5 - default: has credit in default? (categorical: 'no','yes','unknown')<br>
6 - housing: has housing loan? (categorical: 'no','yes','unknown')<br>
7 - loan: has personal loan? (categorical: 'no','yes','unknown')<br>

**related with the last contact of the current campaign:**<br>
8 - contact: contact communication type (categorical: 'cellular','telephone')<br>
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')<br>
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')<br>
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.

**other attributes:**   <br>
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)<br>
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)<br>
14 - previous: number of contacts performed before this campaign and for this client (numeric)<br>
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')<br>

**social and economic context attributes**<br>
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)<br>
17 - cons.price.idx: consumer price index - monthly indicator (numeric)<br>
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)<br>
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)<br>
20 - nr.employed: number of employees - quarterly indicator (numeric)<br>

**Output variable (desired target):**<br>
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')<br>

In [ ]:
# sample of using chaining methods to get the result in differnt lines seprated by . 


# Again Ask Questions and Answer them with Data by Pandas
Q1 : What is the distribution of age(in every decade in data set? <br>
Q2 : Is there relation between age group and balance? <br>
Q3 : Is there relation between marital status and balance? <br>
Q4 : Is there relation between education and balance? <br>
Q5 : Is there relation between occupation and balance? <br>
Q6 : Which job had more default? <br>
Q7 : Which education level had more balance? <br>



In [ ]:
# clean the dataset function


In [ ]:
# get null values


In [ ]:
# clean the dataset function with fillna and assign


In [ ]:
# Answer following questions with the cleaned dataset and chained methods:
bins=[0,20,30,40,50,60,70,80,90,100]
labels=['0-20','20-30','30-40','40-50','50-60','60-70','70-80','80-90','90-100']

(
# Q1 : What is the distribution of age(in every decade in data set? 
    


# Q2 : What is the relation between age group and balance? 


#Q3 : What is the relation between marital status and balance? 



# Q4 : What is the relation between education and balance? 


# Q5 : What is the relation between job and balance? 


# Q6 :  Which job had more default? 


# Q7 : Which education level had more default?


)


### USE QUERY INSIDE PANDAS
when we have a lot of conditions, we can use query method to make our code more readable and efficient. <br>

### Q1: select rows  with balance between 10000 and 20000

In [ ]:
# select rows  with balance between 10000 and 20000
(
)

In [ ]:
# Select rows with balance between 10000 and 20000 and age between 30 and 40 and job is admin or management
(
)

In [ ]:
## Parctice: Select rows with age between 30 and 40 and job is admin or management and had default

#### Simplify Join in Pandas

<img src='img/SQL_JOINS_V2.png'>

#### Concat
#### Merge
#### Join

In [ ]:
# read Bike Sharing Dataset


In [ ]:
# read Calendar Dataset


In [ ]:
# read Weather Dataset


### use Merge to combine two dataframes Then: <br>

Answer the following questions: <br>
Q1 : What is the trend of bike demand over years? <br>
Q2 : Which month(in the whole dataset) had the highest demand for Bikes? <br>
Practice 2 : Is there any relation between temprature and Bike demand? <br>
Practice 3 : Does Holiday affect demand or not? <br>
Practice 4 : Does weathersit affect demand or not? <br>

In [ ]:
# Merge Bike_demand and Calendar_date and Weather_data on date
(
# Q1 : What is the trend of bike demand over years? 


# Which month had the highest demand for Bikes? 

# Practice 2 : Is there any relation between temprature and Bike demand? 

# Practice 3 : Does Holiday affect demand or not 

# Practice 4 : Does weathersit affect demand or not? 

)

### Use concat to combine two dataframes

In [ ]:
# Create a new dataframe for 2018 data


In [ ]:
# Create a new dataframe for 2019 data


In [ ]:
# Concat 2018 and 2019 dataframes into a single dataframe


### Other useful methods
Here I will bring some useful methods that I use them in my daily work. <br>

- isin() -->  isin() method is used to filter data based on a condition. <br>
- str.contains() --> str.contains() method is used to filter data based on a condition. <br>
- np.where() --> np.where() method is used to filter data based on a condition. <br>
- apply() --> apply() method is used to apply a function on a column. <br>
- map() --> map() method is used to map a value to another value. <br>
- unique() --> unique() method is used to get unique values in a column. <br>
- nunique() --> nunique() method is used to get number of unique values in a column. <br>
- value_counts() --> value_counts() method is used to get number of unique values in a column. <br>
- sort_values() --> sort_values() method is used to sort values in a column. <br>
- sort_index() --> sort_index() method is used to sort index in a column. <br>
- drop_duplicates() --> drop_duplicates() method is used to drop duplicates in a column. <br>
- dropna() --> dropna() method is used to drop null values in a column. <br>
- fillna() --> fillna() method is used to fill null values in a column. <br>
- loc,iloc --> loc,iloc method is used to filter data based on a condition. <br>
...

In [ ]:

# Bike_demand_new = (bike.
# .merge(Calendar_date, on='dteday')
# .merge(Weather_data, on='dteday')
# .assign(date = lambda df_: pd.to_datetime(df_.dteday, format='%d-%m-%Y')
# #np.where sample
# ,season_string = lambda df_: np.where(df_.season == 1, 'spring', np.where(df_.season == 2, 'summer', np.where(df_.season == 3, 'fall', 'winter')))
# )
# )
#isin() sample
# select month Jan,feb , march
# Bike_demand_new.loc[Bike_demand_new.date.dt.month.isin((1,2,3)),:]

#unique() sample
# Bike_demand_new.weathersit.unique()

#nunique() sample
# Bike_demand_new.weathersit.nunique()
